File: scripts/add-missing-smart.py
"""
missing-endpoints → manager_tasks xlsx 에 적절한 위치에 행 삽입
- /{id} 항목 제외
- 이전에 하단 추가된 행(Order > 900) 정리
- 각 경로를 같은 seg1 그룹 내 알파벳 순 위치에 삽입
- 삽입된 행: 옅은 빨간색 배경 (#FFE0E0)
"""

import csv, io, os, sys
from copy import copy
from openpyxl import load_workbook
from openpyxl.styles import PatternFill
from openpyxl.utils import get_column_letter

ROOT   = os.path.join(os.path.dirname(__file__), '..')
CSV    = os.path.join(ROOT, 'missing-endpoints.csv')
XLSX   = os.path.join(ROOT, 'manager_tasks_2026-03-12.xlsx')

LIGHT_RED = PatternFill(fill_type='solid', fgColor='FFFFE0E0')  # ARGB: 알파FF + 옅은빨강

# ── CSV 로드 ──
def load_missing():
    entries = []
    with open(CSV, encoding='utf-8') as f:
        reader = csv.reader(f)
        header = next(reader)
        for row in reader:
            if len(row) < 2: continue
            path = row[1].strip()
            if not path: continue
            if path.lower().endswith('/{id}'): continue  # /{id} 제외
            entries.append({
                'tag':     row[0].strip() if len(row) > 0 else '',
                'path':    path,
                'mode':    row[2].strip() if len(row) > 2 else '',
                'methods': row[3].strip() if len(row) > 3 else '',
                'desc':    row[4].strip() if len(row) > 4 else '',
            })
    return entries

def split_path(p):
    parts = p.lstrip('/').split('/')
    parts += ['', '', '', '']
    return parts[0], parts[1], parts[2], parts[3]

# ── 삽입 위치 결정 ──
def find_insert_after(missing_path, missing_tag, existing):
    """
    existing: list of (path, row_idx, tag)  -- row_idx는 1-based
    핵심: 같은 seg1 그룹 내에서 '알파벳으로 가장 가까운 선행 경로'의 행에 삽입.
    즉, path 값이 가장 크면서 missing_path 이하인 항목 → 그 항목의 행 바로 다음.
    반환값: insert_after_row_idx (1-based)
    """
    seg1 = missing_path.split('/')[1].upper() if '/' in missing_path else ''
    missing_upper = missing_path.upper()

    # 1차: 같은 seg1 그룹
    group = [(p, r, t) for p, r, t in existing
             if '/' in p and p.split('/')[1].upper() == seg1]

    # 2차: 같은 Tag
    if not group:
        group = [(p, r, t) for p, r, t in existing if t == missing_tag]

    if not group:
        return None  # 맨 끝에 추가

    # missing_path보다 알파벳 작거나 같은 후보
    candidates = [(p, r) for p, r, t in group if p.upper() <= missing_upper]

    if not candidates:
        # missing_path가 그룹의 첫 항목보다 앞 → 첫 항목 바로 앞에 삽입
        group.sort(key=lambda x: x[1])
        return group[0][1] - 1

    # 후보 중 '알파벳 값이 가장 큰' 경로의 행 선택
    # (= missing_path 직전에 오는 가장 가까운 항목)
    _, best_row = max(candidates, key=lambda x: x[0].upper())
    return best_row

# ── 행 복사 (스타일 유지용) ──
def copy_row_style(ws, src_row, dest_row):
    """src_row의 셀 스타일을 dest_row에 복사"""
    for col in range(1, ws.max_column + 2):
        src_cell = ws.cell(row=src_row, column=col)
        dst_cell = ws.cell(row=dest_row, column=col)
        if src_cell.has_style:
            dst_cell.font      = copy(src_cell.font)
            dst_cell.border    = copy(src_cell.border)
            dst_cell.alignment = copy(src_cell.alignment)

# ── 행 이동 (아래로) ──
def shift_rows_down(ws, from_row, count):
    """from_row 이상의 모든 행을 count만큼 아래로 이동"""
    max_row = ws.max_row
    max_col = ws.max_column

    for row in range(max_row, from_row - 1, -1):
        for col in range(1, max_col + 1):
            src = ws.cell(row=row, column=col)
            dst = ws.cell(row=row + count, column=col)
            dst.value = src.value
            if src.has_style:
                dst.font      = copy(src.font)
                dst.border    = copy(src.border)
                dst.fill      = copy(src.fill)
                dst.number_format = src.number_format
                dst.alignment = copy(src.alignment)
        # 원본 행 지우기
        for col in range(1, max_col + 1):
            cell = ws.cell(row=row, column=col)
            cell.value = None
            cell.fill  = PatternFill(fill_type=None)

# ── 메인 ──
def main():
    print('📂 xlsx 읽는 중...')
    wb = load_workbook(XLSX)
    ws = wb['Tasks']

    # ─ Step 1: 헤더 행 찾기 (Order 컬럼) ─
    header_row = None
    for r in ws.iter_rows():
        if r[0].value == 'Order':
            header_row = r[0].row
            break
    if not header_row:
        print('❌ 헤더 행 없음'); sys.exit(1)
    print(f'헤더 행: {header_row}')

    # ─ Step 2: 이전 추가 행 제거 (Order > 900 또는 빨간 fill) ─
    rows_to_delete = []
    for row in ws.iter_rows(min_row=header_row + 1):
        order_val = row[0].value
        fill_val  = row[0].fill
        is_red = (fill_val and fill_val.patternType == 'solid' and
                  fill_val.fgColor and fill_val.fgColor.type == 'rgb' and
                  fill_val.fgColor.rgb.upper().endswith('FFE0E0'))
        if (isinstance(order_val, (int, float)) and order_val > 900) or is_red:
            rows_to_delete.append(row[0].row)

    # 아래서부터 삭제
    for r in sorted(rows_to_delete, reverse=True):
        ws.delete_rows(r)
    print(f'🗑️  이전 추가 행 {len(rows_to_delete)}개 제거')

    # ─ Step 3: 기존 EndPoint 맵 구성 ─
    existing = []  # (path, row_idx, tag)
    for row in ws.iter_rows(min_row=header_row + 1):
        ep  = row[11].value  # col 12 = index 11
        tag = row[2].value   # col 3  = index 2
        if ep and str(ep).strip():
            existing.append((str(ep).strip(), row[0].row, str(tag).strip() if tag else ''))
    print(f'기존 EndPoint 행: {len(existing)}개')

    # ─ Step 4: missing 로드 및 삽입 계획 ─
    missing = load_missing()
    existing_set = {p.lower() for p, _, _ in existing}

    insert_plan = {}  # after_row_idx -> [entry, ...]
    skip = 0
    for entry in missing:
        if entry['path'].lower() in existing_set:
            skip += 1
            continue
        after = find_insert_after(entry['path'], entry['tag'], existing)
        key = after if after is not None else 'end'
        insert_plan.setdefault(key, []).append(entry)

    total_planned = sum(len(v) for v in insert_plan.values())
    print(f'스킵(기존): {skip}개 / 삽입 예정: {total_planned}개')

    # ─ Step 5: 하단→상단 순으로 삽입 ─
    sorted_keys = sorted(
        insert_plan.keys(),
        key=lambda k: ws.max_row if k == 'end' else k,
        reverse=True
    )

    total_added = 0
    for key in sorted_keys:
        entries = sorted(insert_plan[key], key=lambda e: e['path'])
        insert_at = (ws.max_row + 1) if key == 'end' else key + 1

        for i, entry in enumerate(entries):
            actual_row = insert_at + i
            ws.insert_rows(actual_row)

            seg1, seg2, seg3, seg4 = split_path(entry['path'])

            ws.cell(row=actual_row, column=1).value  = None           # Order
            ws.cell(row=actual_row, column=2).value  = None           # Product
            ws.cell(row=actual_row, column=3).value  = entry['tag']   # Tab
            ws.cell(row=actual_row, column=4).value  = entry['tag']   # Group
            ws.cell(row=actual_row, column=5).value  = entry['desc']  # sub1 = desc
            ws.cell(row=actual_row, column=8).value  = seg1           # seg1
            ws.cell(row=actual_row, column=9).value  = seg2           # seg2
            ws.cell(row=actual_row, column=10).value = seg3           # seg3
            ws.cell(row=actual_row, column=11).value = seg4           # seg4
            ws.cell(row=actual_row, column=12).value = entry['path']  # EndPoint
            ws.cell(row=actual_row, column=13).value = entry['mode']  # mode
            ws.cell(row=actual_row, column=22).value = entry['methods'] # remark

            # 빨간 배경
            for col in range(1, 25):
                ws.cell(row=actual_row, column=col).fill = LIGHT_RED

            total_added += 1

    print(f'\n✅ 총 {total_added}개 행 삽입 완료')

    wb.save(XLSX)
    print(f'💾 저장: {XLSX}')

    # ─ 검증: MVLD 근처 ─
    print('\n📋 검증 - MVLD 관련 행:')
    wb2 = load_workbook(XLSX)
    ws2 = wb2['Tasks']
    for row in ws2.iter_rows(min_row=3):
        ep = row[11].value
        if ep and 'MVLD' in str(ep).upper():
            fill = row[0].fill
            is_new = (fill and fill.fgColor and fill.fgColor.type == 'rgb'
                      and fill.fgColor.rgb in ('FFFFE0E0', 'FFE0E0'))
            marker = ' [NEW🔴]' if is_new else ''
            print(f'  Row{row[0].row}  EP:{ep}{marker}')

if __name__ == '__main__':
    main()

File: scripts/add-missing-smart.mjs
/**
 * missing-endpoints → manager_tasks xlsx 적합한 위치에 행 삽입
 * 1. 이전 스크립트가 하단에 추가한 rows 제거 (Order > 900)
 * 2. 각 엔드포인트를 알파벳 순 기준으로 같은 seg1 그룹 내 적절한 위치에 삽입
 * 3. 삽입된 행: 옅은 빨간색 배경 (#FFE0E0)
 */

import fs from 'fs';
import path from 'path';
import { fileURLToPath } from 'url';
import ExcelJS from 'exceljs';

const __dirname = path.dirname(fileURLToPath(import.meta.url));
const ROOT = path.join(__dirname, '..');
const CSV_PATH  = path.join(ROOT, 'missing-endpoints.csv');
const XLSX_PATH = path.join(ROOT, 'manager_tasks_2026-03-12.xlsx');

const LIGHT_RED = { type: 'pattern', pattern: 'solid', fgColor: { argb: 'FFFFE0E0' } };

// ── CSV 파싱 ─────────────────────────────────────────────────
function parseCSVLine(line) {
  const cols = [];
  let inQ = false, cur = '';
  for (let i = 0; i < line.length; i++) {
    const c = line[i];
    if (c === '"') {
      if (inQ && line[i + 1] === '"') { cur += '"'; i++; }
      else inQ = !inQ;
    } else if (c === ',' && !inQ) {
      cols.push(cur); cur = '';
    } else { cur += c; }
  }
  cols.push(cur);
  return cols;
}

function loadMissingEndpoints() {
  const raw = fs.readFileSync(CSV_PATH, 'utf-8');
  const lines = raw.split(/\r?\n/).filter(l => l.trim());
  const entries = [];
  for (let i = 1; i < lines.length; i++) {
    const cols = parseCSVLine(lines[i]);
    const pathUrl = (cols[1] || '').trim();
    if (!pathUrl) continue;
    // /{id} 제외
    if (pathUrl.toLowerCase().endsWith('/{id}')) continue;
    entries.push({
      tag:     (cols[0] || '').trim(),
      path:    pathUrl,
      mode:    (cols[2] || '').trim(),
      methods: (cols[3] || '').trim(),
      desc:    (cols[4] || '').trim(),
    });
  }
  return entries;
}

function splitPath(urlPath) {
  const parts = urlPath.replace(/^\//, '').split('/');
  return [parts[0]||'', parts[1]||'', parts[2]||'', parts[3]||''];
}

// ── 삽입 위치 결정 ────────────────────────────────────────────
// existingEntries: [{path, rowNum, tag}] (rowNum은 현재 실제 sheet row)
function findInsertAfterRow(missingPath, missingTag, existingEntries) {
  const seg1 = missingPath.split('/')[1] || '';
  const missingUpper = missingPath.toUpperCase();

  // 1차: 같은 seg1 그룹 (대소문자 무시)
  let group = existingEntries.filter(e =>
    (e.path.split('/')[1] || '').toUpperCase() === seg1.toUpperCase()
  );

  // seg1 그룹이 없으면 2차: 같은 Tag
  if (group.length === 0) {
    group = existingEntries.filter(e => e.tag === missingTag);
  }

  if (group.length === 0) return -1; // 찾지 못함 → 맨 끝에 추가

  // 그룹 내 마지막 row (알파벳 기준 포지션 찾기)
  // 그룹을 rowNum 순으로 정렬
  group.sort((a, b) => a.rowNum - b.rowNum);

  // path <= missingPath 인 마지막 항목 → insert after
  let bestRowNum = group[0].rowNum - 1; // 그룹 시작 전 (최악의 경우 그룹 앞)
  for (const e of group) {
    if (e.path.toUpperCase() <= missingUpper) {
      bestRowNum = e.rowNum;
    }
  }
  return bestRowNum;
}

// ── 메인 ─────────────────────────────────────────────────────
async function main() {
  console.log('📂 xlsx 읽는 중...');
  const wb = new ExcelJS.Workbook();
  await wb.xlsx.readFile(XLSX_PATH);
  const ws = wb.getWorksheet('Tasks');
  if (!ws) { console.error('Tasks 시트 없음'); process.exit(1); }

  // ─ Step 1: 이전에 추가된 하단 행 제거 (Order > 900) ─
  const rowsToDelete = [];
  ws.eachRow((row, rowNum) => {
    if (rowNum <= 2) return;
    const v = row.getCell(1).value;
    if (typeof v === 'number' && v > 900) rowsToDelete.push(rowNum);
    // Order가 없더라도 light-red fill인 행 (이전 추가분 일부)
    else {
      const fill = row.getCell(1).fill;
      if (fill && fill.fgColor && fill.fgColor.argb === 'FFFFE0E0') {
        rowsToDelete.push(rowNum);
      }
    }
  });
  rowsToDelete.sort((a, b) => b - a);
  for (const r of rowsToDelete) ws.spliceRows(r, 1);
  console.log(`🗑️  이전 추가 행 ${rowsToDelete.length}개 제거`);

  // ─ Step 2: 현재 엔드포인트 맵 구성 ─
  const existingEntries = [];
  ws.eachRow((row, rowNum) => {
    if (rowNum <= 2) return;
    const ep = row.getCell(12).value;
    const tag = row.getCell(3).value;
    if (ep && String(ep).trim()) {
      existingEntries.push({
        path:   String(ep).trim(),
        rowNum,
        tag:    tag ? String(tag).trim() : '',
      });
    }
  });
  console.log(`기존 EndPoint 행: ${existingEntries.length}개`);

  // ─ Step 3: missing 엔드포인트 로드 ─
  const missing = loadMissingEndpoints();
  const existingSet = new Set(existingEntries.map(e => e.path.toLowerCase()));

  // ─ Step 4: 삽입 계획 수립 ─
  // insertPlan: Map<afterRowNum, entry[]>
  const insertPlan = new Map();
  let skipCount = 0;

  for (const entry of missing) {
    if (existingSet.has(entry.path.toLowerCase())) { skipCount++; continue; }

    const afterRow = findInsertAfterRow(entry.path, entry.tag, existingEntries);
    const key = afterRow === -1 ? 99999 : afterRow; // -1이면 맨 끝
    if (!insertPlan.has(key)) insertPlan.set(key, []);
    insertPlan.get(key).push(entry);
  }

  console.log(`스킵(기존): ${skipCount}개 / 삽입 예정: ${[...insertPlan.values()].flat().length}개`);

  // ─ Step 5: 하단→상단 순으로 삽입 ─
  const sortedPositions = [...insertPlan.keys()].sort((a, b) => b - a);
  let totalAdded = 0;

  for (const afterRowNum of sortedPositions) {
    const entries = insertPlan.get(afterRowNum);
    // 각 position에서 알파벳 순 정렬
    entries.sort((a, b) => a.path.localeCompare(b.path));

    // spliceRows로 한 번에 삽입
    let lastRowNum = 2;
    ws.eachRow((r, n) => { lastRowNum = Math.max(lastRowNum, n); });
    const insertAt = afterRowNum === 99999
      ? lastRowNum + 1
      : afterRowNum + 1;

    const rowArrays = entries.map(entry => {
      const [seg1, seg2, seg3, seg4] = splitPath(entry.path);
      const arr = Array(28).fill(null);
      arr[0]  = null;          // Order (비워둠 - 나중에 수동 채움)
      arr[1]  = null;          // Product
      arr[2]  = entry.tag;     // Tab
      arr[3]  = entry.tag;     // Group
      arr[4]  = entry.desc;    // sub1 (description)
      arr[5]  = null;
      arr[6]  = null;
      arr[7]  = seg1;          // seg1
      arr[8]  = seg2;          // seg2
      arr[9]  = seg3;          // seg3
      arr[10] = seg4;          // seg4
      arr[11] = entry.path;    // EndPoint
      arr[12] = entry.mode;    // mode
      arr[21] = entry.methods; // remark (col 22, idx 21)
      return arr;
    });

    ws.spliceRows(insertAt, 0, ...rowArrays);

    // 삽입된 행에 빨간색 배경 적용
    for (let i = 0; i < entries.length; i++) {
      const row = ws.getRow(insertAt + i);
      for (let col = 1; col <= 24; col++) {
        const cell = row.getCell(col);
        cell.fill = LIGHT_RED;
      }
      row.commit();
    }

    totalAdded += entries.length;
  }

  console.log(`\n✅ 총 ${totalAdded}개 행 삽입 완료`);

  // ─ 저장 ─
  await wb.xlsx.writeFile(XLSX_PATH);
  console.log(`💾 저장: ${XLSX_PATH}`);

  // ─ 검증: MVLD 근처 확인 ─
  console.log('\n📋 검증 - MVLD 근처 행:');
  const wb2 = new ExcelJS.Workbook();
  await wb2.xlsx.readFile(XLSX_PATH);
  const ws2 = wb2.getWorksheet('Tasks');
  ws2.eachRow((row, rowNum) => {
    const ep = row.getCell(12).value;
    if (ep && String(ep).toUpperCase().includes('MVLD')) {
      const fill = row.getCell(1).fill;
      const isNew = fill && fill.fgColor && fill.fgColor.argb === 'FFFFE0E0' ? '[NEW]' : '';
      console.log(`  Row${rowNum} EP:${ep} ${isNew}`);
    }
  });
}

main().catch(err => { console.error('❌', err); process.exit(1); });
